### PR TITLE
#1513 LLPanelPreferenceGameControl: Implement UI for device settings

### PR DESCRIPTION
**Pull request by LLGuru**

### CODE REVIEW

File: indra/llcommon/llinitparam.h [hunk 1]
```diff
@@ -1772,8 +1772,13 @@ namespace LLInitParam
 
 		// implicit conversion
 		operator const container_t&() const { return mValues; } 
+		operator container_t&() { return mValues; } 
 		// explicit conversion
 		const container_t& operator()() const { return mValues; }
+		container_t& operator()() { return mValues; }
+		// direct Nth item access
+		const typename NAME_VALUE_LOOKUP::type_value_t& operator()(size_t index) const { return mValues[index]; }
+		typename NAME_VALUE_LOOKUP::type_value_t& operator()(size_t index) { return mValues[index]; }
```
Comment (marchcat) [inline indra/llcommon/llinitparam.h:1781]:
```suggestion
    const typename NAME_VALUE_LOOKUP::type_value_t& operator()(size_t index) const
    {
        llassert(index < mValues.size());
        return mValues[index];
    }
    
    typename NAME_VALUE_LOOKUP::type_value_t& operator()(size_t index)
    {
        llassert(index < mValues.size());
        return mValues[index];
    }
```

File: indra/newview/llfloaterpreference.cpp [hunk 2]
```diff
@@ -3290,18 +3409,46 @@ BOOL LLPanelPreferenceGameControl::postBuild()
...
+    populateMappingTableRows(mButtonMappings, mBinaryChannelSelector, LLGameControl::NUM_BUTTONS);
+
+    // Workaround for the common bug:
+    // LLScrollListCtrl with draw_heading="true" initially has incorrect mTop (17 px higher)
+    LLRect rect = mAxisOptions->getRect();
+    rect.mTop = mAxisOptions->getParent()->getRect().getHeight() - 1;
+    mAxisOptions->setRect(rect);
+    mAxisOptions->updateLayout();
+
     return TRUE;
```
Comment (marchcat) [inline indra/newview/llfloaterpreference.cpp:3452]: Might be good to use `bool` here.

Comment (LLGuru) [reply indra/newview/llfloaterpreference.cpp:3452]: Later, after merge with main

Comment (marchcat) [reply indra/newview/llfloaterpreference.cpp:3452]: But on a flip side, let's do it within a separate project later.

File: indra/llcommon/llinitparam.h (hunk 1 above)
Comment (LLGuru) [reply indra/llcommon/llinitparam.h:1781]: No needs for these additional chacks

**Review by marchcat — APPROVED** [review]

File: indra/llwindow/llgamecontrol.cpp [hunk 3]
```diff
@@ -1398,8 +1399,8 @@ void onControllerDeviceAdded(const SDL_Event& event)
 {
     std::string guid(std::to_string(SDL_JoystickGetDeviceGUID(event.cdevice.which)));
     SDL_GameControllerType type(SDL_GameControllerTypeForIndex(event.cdevice.which));
-    std::string name(SDL_GameControllerNameForIndex(event.cdevice.which));
-    std::string path(SDL_GameControllerPathForIndex(event.cdevice.which));
+    std::string name(std::to_string(SDL_GameControllerNameForIndex(event.cdevice.which)));
+    std::string path(std::to_string(SDL_GameControllerPathForIndex(event.cdevice.which)));
```
Comment (AndrewMeadows) [inline indra/llwindow/llgamecontrol.cpp:1403]: I'm not seeing an version of `std::to_string()` in the documentation that accepts a `const char*` as an argument.  Did we provide a custom override that handles the `nullptr` or... how does this work?

**Review by AndrewMeadows — CHANGES_REQUESTED** [review]
Remove the useless `if() return `around `function_call_with_side_effects()`.

File: indra/llwindow/llgamecontrol.cpp [hunk 4]
```diff
@@ -1887,7 +1884,7 @@ bool LLGameControl::parseDeviceOptions(const std::string& options, std::string&
         if (!value.empty())
         {
             size_t number = std::stoull(value);
-            if (number > MAX_BUTTON || std::to_string(number) != value)
+            if (number > NUM_AXES || std::to_string(number) != value)
```
[the remark this answers is not in the document]
Comment (LLGuru) [reply indra/llwindow/llgamecontrol.cpp]: Fixed

File: indra/newview/llfloaterpreference.cpp [hunk 5]
```diff
@@ -3258,13 +3310,77 @@ void LLPanelPreferenceGameControl::applyGameControlInput()
...
+            {
+                mNumericValueEditor->setMinValue(-LLGameControl::MAX_AXIS_OFFSET);
+                mNumericValueEditor->setMaxValue(LLGameControl::MAX_AXIS_OFFSET);
+                mNumericValueEditor->setValue(deviceOptions.getAxisOptions()[row_index].mOffset);
+            }
+            mNumericValueEditor->setVisible(TRUE);
+        }
+
+        if (initCombobox(row, mAxisOptions))
+            return;
```
[the remark this answers is not in the document]
Comment (LLGuru) [reply indra/newview/llfloaterpreference.cpp:3354]: Fixed

File: indra/newview/llfloaterpreference.cpp [hunk 6]
```diff
@@ -3167,62 +3171,110 @@ bool LLPanelPreferenceGameControl::initChannelSelector(LLScrollListItem* item)
...
+    else if (grid == mAxisMappings)
+    {
+        combobox = mAxisSelector;
+    }
+    else if (grid == mButtonMappings)
+    {
+        combobox = mBinaryChannelSelector;
+    }
+    if (!combobox)
         return false;
```
[the remark this answers is not in the document]
Comment (LLGuru) [reply indra/newview/llfloaterpreference.cpp:3194]: Not agree
The first clause of `if` could set `combobox` to `null` - in this case we also should return

File: indra/llwindow/llgamecontrol.cpp (hunk 3 above)
Comment (LLGuru) [reply indra/llwindow/llgamecontrol.cpp:1403]: Yes

**Comment by LLGuru** [comment]
> Remove the useless `if() return `around `function_call_with_side_effects()`.

Done:
https://github.com/secondlife/viewer/blob/4eb24ba89c4560cec99d254cf98ef25c4e7648b5/indra/newview/llfloaterpreference.cpp#L3352-L3354

Approve?

**Pull request merged**